std/digest/murmurhash.d

```diff
@@ -0,0 +1,860 @@
+    }
+
+    /// Returns the current hashed value as an ubyte array.
+    auto getBytes() pure nothrow @nogc
+    {
+        return cast(ubyte[Block.sizeof]) get();
+    }
+}
+
+unittest
+{
+    // dfmt off
+    checkResult!FastMurmurHash3_128_opt64([
+        "" : "00000000000000000000000000000000",
+        "a" : "897859F6655555855A890E51483AB5E6",
+        "ab" : "2E1BED16EA118B93ADD4529B01A75EE6",
+        "abc" : "6778AD3F3F3F96B4522DCA264174A23B",
+        "abcd" : "4FCD5646D6B77BB875E87360883E00F2",
+        "abcde" : "B8BB96F491D036208CECCF4BA0EEC7C5",
+        "abcdef" : "55BFA3ACBF867DE45C842133990971B0",
+        "abcdefg" : "99E49EC09F2FCDA6B6BB55B13AA23A1C",
+        "abcdefgh" : "028CEF37B00A8ACCA14069EB600D8948",
+        "abcdefghi" : "64793CF1CFC0470533E041B7F53DB579",
+        "abcdefghij" : "998C2F770D5BC1B6C91A658CDC854DA2",
+        "abcdefghijk" : "029D78DFB8D095A871E75A45E2317CBB",
+        "abcdefghijkl" : "94E17AE6B19BF38E1C62FF7232309E1F",
+        "abcdefghijklm" : "73FAC0A78D2848167FCCE70DFF7B652E",
+        "abcdefghijklmn" : "E075C3F5A794D09124336AD2276009EE",
+        "abcdefghijklmno" : "FB2F0C895124BE8A612A969C2D8C546A",
+        "abcdefghijklmnop" : "23B74C22A33CCAC41AEB31B395D63343",
+        "abcdefghijklmnopq" : "57A6BD887F746475E40D11A19D49DAEC",
+        "abcdefghijklmnopqr" : "508A7F90EC8CF0776BC7005A29A8D471",
+        "abcdefghijklmnopqrs" : "886D9EDE23BC901574946FB62A4D8AA6",
+        "abcdefghijklmnopqrst" : "F1E237F926370B314BD016572AF40996",
+        "abcdefghijklmnopqrstu" : "3CC9FF79E268D5C9FB3C9BE9C148CCD7",
+        "abcdefghijklmnopqrstuv" : "56F8ABF430E388956DA9F4A8741FDB46",
+        "abcdefghijklmnopqrstuvw" : "8E234F9DBA0A4840FFE9541CEBB7BE83",
+        "abcdefghijklmnopqrstuvwx" : "F72CDED40F96946408F22153A3CF0F79",
+        "abcdefghijklmnopqrstuvwxy" : "0F96072FA4CBE771DBBD9E398115EEED",
+        "abcdefghijklmnopqrstuvwxyz" : "A94A6F517E9D9C7429D5A7B6899CADE9"]);
+    // dfmt on
+}
+
+unittest
+{
+    // Pushing unaligned data and making sure the result is still coherent.
+    void testUnalignedHash(H)()
+    {
+        immutable ubyte[1025] data = 0xAC;
+        immutable alignedHash = digest!H(data[0 .. $ - 1]); // 0..1023
+        immutable unalignedHash = digest!H(data[1 .. $]); // 1..1024
+        assert(alignedHash == unalignedHash);
+    }
+
+    testUnalignedHash!MurmurHash3_32_opt32();
+    testUnalignedHash!MurmurHash3_128_opt32();
+    testUnalignedHash!MurmurHash3_128_opt64();
+}
+
+import std.traits : moduleName;
+
+/*
+This is a helper struct and is not intended to be used directly. MurmurHash
+cannot put chunks smaller than Block.sizeof at a time. This struct stores
+remainder bytes in a buffer and pushes it when the block is complete or during
+finalization.
+*/
+struct Piecewise(Hasher) if (moduleName!Hasher == "std.digest.murmurhash")
+{
+    enum blockSize = bits!Block;
+
+    alias Block = Hasher.Block;
+    union BufferUnion
```

Comment:
Because of the `union` you'd need to add the occasional `@trusted` to the users of this structure.